### PR TITLE
Cython ND Array iterator

### PR DESCRIPTION
**Pull request by vighneshbirodkar**
Thoughts appreciated.
Will submit the neighborhood iterator once this is approved.

### CODE REVIEW

**Comment by jni** [comment]
What's the difference between pxi and pyx?

File: skimage/util/_iterator_tests.pyx [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    cdef int sum
+
+    sum = 0
+
+    while iter.has_next():
+        sum += iter.get_int()
+        iter.next()
+
+    del iter
+    assert np.sum(arr) == sum
```
Comment (jni) [inline skimage/util/_iterator_tests.pyx]: Don't you need to import numpy somewhere? Also, don't overwrite built-in names...

File: skimage/util/_iterator_tests.pyx [hunk 2]
```diff
@@ -0,0 +1,17 @@
+include "npy_iterator.pxi"
+
+
+def run_iterator_tests():
+    arr = np.arange(720).reshape(6, 5, 4, 3, 2).astype(np.int)
+
+    cdef Iterator iter = Iterator(arr)
```
Comment (jni) [inline skimage/util/_iterator_tests.pyx]: (This is probably related to my question about importing numpy...) I don't know how pure-C imports work in Cython, but isn't there a way to make them "scoped" by namespace as in Python?

**Comment by jni** [comment]
Again, maybe my Cython noob-ness showing, but is it possible to make this more Pythonic? Here's how I would like it to read:

```python
while iter:
    total += next(iter)
```

I particularly dislike the `get_int` and `get_float` methods... imho the iterator should "emit" the same type as the underlying array.

Finally, quite often, we want not just the value but the current index. Is there a way for you to get that easily with a `current_index` method? (This will be necessary for the neighbour iterator anyway, I think.)

**Comment by vighneshbirodkar** [comment]
@jni, a `pyx` file is to be compiled into C code and eventually an `.so` or `.dll` file. A `pxi` is analogous to the C `#include` directive, it will simply copy paste the contents.

**Comment by njsmith** [comment]
I am summoned! Uh. What am I looking at, and why?

**Comment by ahojnnes** [comment]
I probably missed the discussion, but how is this useful or better than `for item in iterator`?

**Comment by vighneshbirodkar** [comment]
@ahojones, this translates to a much more efficient C code and is intended for fadt Cython ND iteration. See #1495

@njsmith We are trying to write a ease to use as well as fast l Cython wrapper around tge NumPy iterator API

**Comment by jni** [comment]
@vighneshbirodkar is the `.pxi` significantly faster? As I mentioned, the lack of scoping is really frustrating to me...

**Comment by stefanv** [comment]
I think they should be .pxd, not .pxi?

**Comment by ahojnnes** [comment]
Why is this not a PR to Cython then? It should be relatively easy to generate the appropriate code whenever the iterable is a numpy array? So, that you can actually use nice syntax, e.g. `for i in array`...

**Comment by njsmith** [comment]
This looks really weird to me. Aside from the part where you're writing a .pxi instead of a .pxd... you can't `free` the nditer object, it's a regular refcounted Python object. And the exposed interface (of doing a single one-element-at-a-time iteration over a single array) is incredibly tiny compared to what nditer already exposes to C and Python. And the methods are all like single-line renames of existing nditer functions? (I guess I could be confused and you're wrapping one of the older iteration interfaces in numpy, but I'm 99% sure that `PyArrayIterObject` is the same as nditer.)

Like, if you just call the Python function `np.nditer` (which has a very rich construction API and who cares if it's a bit slow b/c it's not the inner loop), this returns a new Python object, and you just have to tell Cython that btw this Python object is specifically of the `PyArrayIterObject` type, and now you can call the low-level stuff on it directly while letting Cython take care of refcnting. Wouldn't that make more sense?

**Comment by vighneshbirodkar** [comment]
Hello @njsmith , @stefanv  thanks for looking into this
According to the [numpy pxd file](https://github.com/cython/cython/blob/master/Cython/Includes/numpy/__init__.pxd#L194), keeping function definitions in `.pxd` file is still experimental, that's why I decided to write a `.pxi` file. With that said, I'll give it a try and see what happens. I will remove the `free` call.

Yes, the functionality offered by my code is indeed trivial. My only intent is to let it's user not have to do pointer manipulation and C imports in order to iterate over an ND array.  With this, all he has to do is call `next` and `get_*` functions.

As you said `np.nditer` would do this with some loss of speed. But the problem comes when we try to iterate over an ND array and its ND neighborhood. This would need 2 python iterators one inside the other, that's when things start getting really slow. That's why I eventually also plan to wrap the [Neighborhood Iterator](http://docs.scipy.org/doc/numpy/reference/c-api.array.html#neighborhood-iterator). I believe iterating over an ND image and its ND neighborhood is pretty useful functionality to have. You can do it now using the C API from withing Cython, but that would make the code a lot less Pythonic and shift your focus to getting the imports and pointer manipulation right.

TL;DR The API(when complete) will let the user iterate over an ND array and its neighborhood without being familiar with Numpy C API.

**Comment by njsmith** [comment]
If you just remove the free() then you'll have a memory leak. You need to either call Py_D
ECREF or (better) use the syntax that tells cython that these objects are actually python objects and not just C structs, so that it does refcnting automatically. See how PyArrayObject gets wrapped in numpy.pxd.

> With this, all he has to do is call next and get_* functions.

Friendly reminder that not all programmers are male.

> As you said np.nditer would do this with some loss of speed.

You're using nditer, just via a really limited api. The loss of speed of using the full api would be unmeasurable for any reasonable sized image and would make your code simpler -- all you really need for your purposes is some utility functions (not a class) that call next and return properly cast pointers. For the neighborhood iterator then yeah iirc there is currently no python wrapper for the constructor function so you also need to wrap that. (Maybe numpy should have a python level api for this built in though, if you get inspired... ;-))

**Comment by vighneshbirodkar** [comment]
@njsmith Now that you point it out, the class does seem like an overkill, will make the changes soon. 

I would disagree with you regarding the loss of speed. Python iterators would make python function calls which will come with its own over head. On the other hand `PyArray_ITER_NEXT` is just a macro with a for loop inside.

**Comment by vighneshbirodkar** [comment]
@njsmith I'll run a quick benchmark and share the result soon.

**Comment by vighneshbirodkar** [comment]
@njsmith See https://github.com/vighneshbirodkar/cy_iter/blob/master/iter.pyx

The code is not exactly the same, but similar, the results on my system were
```shell
Numpy took  0.000637054443359
ans =  2161916
Cython took  0.00178599357605
ans =  2161916
Python took  0.204711914062
ans =  2161916
```
I am assuming that the eventual operation intended is not something as simply as the arithmetic sum.

Now think about iterating over neighbors, let's say over a 2D image. For each pixel, you would have a Python iterator going over 8 values. The difference would be even more obvious then.

**Comment by njsmith** [comment]
Read again, you're seeing what you expect me to say, not what I'm saying
:-). You can call PyIter_ITER_NEXT on the object returned by np.nditer

**Comment by vighneshbirodkar** [comment]
@njsmith 
Okay :-). I'll re-factor the file and break it into functions.

Also, multiple `get_*` functions are unattractive, as Juan also pointed out. Any idea how we could use fused types to solve this ? My question on the Cython mailing list is yet unanswered :-( 
https://groups.google.com/forum/#!topic/cython-users/rcS4ZRXXkgM

**Comment by vighneshbirodkar** [comment]
@njsmith I am not wrong the `nditer` function relies on the newer `NpyIter` struct. This eventually uses the [NpyIter_GetIterNext](http://docs.scipy.org/doc/numpy/reference/c-api.iterator.html#c.NpyIter_GetIterNext) function which returns a function pointer to be used for iteration. Wouldn't is be more convenient if this were wrapped up in a class ?

**Comment by njsmith** [comment]
Ah, yeah, you're probably right :-). I am not at all an expert on the details of numpy's iteration APIs (and they're very confusing!), so while I tried to figure out what exactly was going on with `PyArrayIterObject` before posting I probably got it wrong.

I can't really say whether any particular API would be more or less convenient than just using the existing numpy API directly. If the only difference is that the new API is better documented then you're probably better off submitting doc patches :-). If you have other improvement ideas then we'd certainly like to hear them upstream too, though that wouldn't help in the short term (b/c skimage needs to support older numpys).

**Comment by JDWarner** [comment]
Just my 2 cents - I have no problem with Cython up through typed memoryviews, but find the NumPy C API difficult to grok. Even if it's possible using pointers and the C API presently, there's a reason few of our algorithms go this route right now - it's a barrier both to implementation and maintenance. An efficient and elegant Cython implementation, even if just a wrapper (and even if slightly less efficient), would let us realize these gains more generally across the package.

**Comment by vighneshbirodkar** [comment]
Since we support numpy 0.9, we can't assume the Npy_* API available to us.
And, as @njsmith suggested,  we cannot be sure `np.nditer` wraps which one
of these API.

And, as the documentation[1] suggests, neighbourhood iteration is not
supported with the new API.

I'll go ahead and refactor this class into utility functions using the same
iterator API.

[1] http://docs.scipy.org/doc/numpy/reference/c-api.iterator.html

**Comment by jaimefrio** [comment]
Hi Vighnesh,

You may want to take a look at the Cython implementation of labeling in `ndimage` [here](https://github.com/scipy/scipy/blob/master/scipy/ndimage/src/_ni_label.pyx). It uses `PyArrayIterObject` from Cython, and even has some examples of fused data type specialization to copy cast data from the iteration point into a contiguous buffer. A lot of it is just a convoluted mess of function pointers, that would be infinitely more clearly written in C directly, but I digress...

Anyway, @njsmith pointed you in the wrong direction, but with good intentions. ;-) `PyArrayIterObject` is indeed exposed as a Python object, only it is not `np.nditer`, but `np.flatiter`. As Nathaniel pointed out, you can call `PyArray_ITER_NEXT` on the return of `np.flatiter`. There are some issues with the way yu are accessing the data in the array, but I'll comment on that inline.

As you mentioned elsewhere, this is highly overlapping with @aman-iitj and his GSoC project. This said, I by no means want to push you away from working on this, but rather pull you in so that your work is coordinated with Aman's.

File: skimage/util/npy_iterator.pxi [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        """ Points the iterator to the next element of the array.
+
+        The behaviour of this function when `has_next` is `False`, is undefined
+        and may result in a crash.
+        """
+        PyArray_ITER_NEXT(self.iter)
+
+    cdef inline cnp.float_t get_float(self):
+        """ Returns the current element as `float`. """
+        cdef cnp.float_t *address = <cnp.float_t*>self.iter.dataptr
```
Comment (jaimefrio) [inline skimage/util/npy_iterator.pxi]: Unfortunately you cannot do this... At least not without previously checking that the array being iterated is aligned and its type not byteswapped. The dtype object has a pointer to a `copyswap` function, see [nere](http://docs.scipy.org/doc/numpy/reference/c-api.types-and-structures.html#c.PyArray_ArrFuncs.copyswap), that can be sued to copy data into an aligned buffer which you can then dereference without worries.

**Comment by stefanv** [comment]
@jaimefrio If it can be written more clearly in C, shouldn't we do that and wrap the C with Cython?

File: skimage/util/npy_iterator.pxi (hunk 3 above)
Comment (vighneshbirodkar) [inline skimage/util/npy_iterator.pxi]: Hello @jaimefrio 
I plan to re visit this soon
Could you elaborate what you mean by aligned ? Or could you point me to some documentation which elaborates it ?

And I am guessing here by byteswapped you mean I should check for the endianness of the array ?

**Comment by matthew-brett** [comment]
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
